### ISSUE TITLE
[Bug]: docs-mode table-of-contents does not change URL when clicking on an anchor

### ISSUE THREAD

**Issue by kroeder** [storybookjs/storybook#26345]
### Describe the bug

If you enable `toc: true`, docs-mode renders a table-of-contents inside the docs page.
When clicking on one of the toc list-items, the expected behavior would be to add the # anchor to the browser URL.

It should behave the same as clicking the # icon on a docs-page headline.

### To Reproduce

https://stackblitz.com/edit/github-w91kck?file=.storybook%2Fpreview.ts

- Open "Button / Docs"
- Click on table of contents anchor -> nothing is added to the URL

**Comment by vanessayuenn** [comment]
@kroeder in the repro you provided, the TOC does link to the # anchor. Can you clarify how do you expect it to work differently?

**Comment by ftr93** [comment]
@vanessayuenn yes, it links to the anchor correctly, but clicking it doesn't change the browser URL

**Comment by sooster910** [comment]
I have confirmed the behavior @kroeder described: When hovering over the anchor tag, as shown in the image below, it behaves as expected. However, upon clicking, the browser URL does not update to reflect the clicked anchor tag.

<img width="492" alt="image" src="https://github.com/storybookjs/storybook/assets/26635607/0da7c660-ea50-4231-90fc-e672f66572a0">

@vanessayuenn  can I give it a try? Thanks.

**Comment by vanessayuenn** [comment]
@sooster910 of course! We'd be thrilled to have your contribution. If this is your first contribution, be sure to check the [contribution guide](https://storybook.js.org/docs/react/contribute/how-to-contribute) and hop onto the [Storybook Discord](https://discord.gg/storybook) if you need any help getting started. Looking forward to your PR! ✨

**Comment by emlynmac** [comment]
Just came across this issue in a repo I'm working with.

@sooster910 can you update your PR to be conflict-free? 
@vanessayuenn would you be able to review once the updates are made?

**Comment by sooster910** [comment]
> Just came across this issue in a repo I'm working with.
> 
> [@sooster910](https://github.com/sooster910) can you update your PR to no conflicts? [@vanessayuenn](https://github.com/vanessayuenn) would you be able to review once the updates are made?

@emlynmac  Sure. Given the conflicts with the original PR, I'll go ahead and create a new PR to address the issue from scratch.

@vanessayuenn once the new PR is up, I would appreciate it if you could review it.  Thanks.

**Comment by kroeder** [comment]
This could be related https://github.com/storybookjs/storybook/issues/29361

If you have an anchor hash value, 

1. On page load, it gets removed from the story (which would make sharing the hash value of a toc-entry useless)
2. When adding an anchor hash value, the page does not scroll to the anchor

@sooster910 by any chance, did your PR fix problem 2? If not, that's fine! I'm currently investigating https://github.com/storybookjs/storybook/issues/29361 myself

**Comment by sooster910** [comment]
> This could be related [#29361](https://github.com/storybookjs/storybook/issues/29361)
> 
> If you have an anchor hash value,
> 
> 1. On page load, it gets removed from the URL (which would make sharing the hash value of a toc-entry useless)
> 2. When adding an anchor hash value, the page does not scroll to the anchor
> 
> [@sooster910](https://github.com/sooster910) by any chance, did your PR fix problem 2? If not, that's fine! I'm currently investigating [#29361](https://github.com/storybookjs/storybook/issues/29361) myself

@kroeder   hi, the scrolling only works when you click the TOC now - the click handler updates the anchor, but if you type the anchor directly in the address bar, it doesn't work. Let me know if you need any clarification!

**Comment by Sidnioulz** [comment]
https://github.com/storybookjs/storybook/pull/30130/files should fix this based on an initial PR from @sookmax.

I'd love for someone in this thread to check out that PR and confirm the bug is fixed for them.
